parser.js

```diff
@@ -144,6 +144,20 @@
 
 (function() {
 
+  // Simplified implementation of String.fromCodePoint for legacy environments.
+  // This function doesn't check that the parameter is a positive integer
+  // and <= 0x10ffff, since it will only be called with valid codepoints.
+  var fromCodePoint = String.fromCodePoint || function(codepoint) {
+    if (codepoint <= 0xffff) { // BMP
+      return String.fromCharCode(codepoint);
+    } else { // astral
+      var cp = codepoint - 0x10000;
+      var highSurrogate = (cp >> 10) + 0xd800;
+      var lowSurrogate = (cp % 1024) + 0xdc00;
+      return String.fromCharCode(highSurrogate, lowSurrogate);
+    }
+  };
```

Comment:
Two questions:

1) Where did you get this code from? If you copied it, is the copied code license compatible with the license used in this repo?

2) Do you think there is value in using a more evolved polyfill like the one provided at [1]

[1]: https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/String/fromCodePoint

Reply:
1. I copied i from a PR I opened some years ago (https://github.com/jshint/jshint/pull/2413).
  a. I *think* I wrote it, because I couldn't find a similar implementation on the web (searching for code is hard lol, I looked at the polyfills I could find and searched `"lowSurrogate" "highSurrogate"` on Google).
  b. The JSHint license is MIT*
    <details>
    <summary>*</summary>

    Expect for the original code from JSLint, for which is the license is JSON, but it doesn't affect that PR
    </details>

2. The polyfill on MDN is more spec-compliant, since it works with multiple parameters and throws on invalid codepoints. As noted in the comment I wrote before this function, it is simplified to only handle the parameters that `regjsparser` could call it with. If you prefer I can use a more complete polyfill, but I don't think that it is necessary.

Reply:
Thx for the reply!

> If you prefer I can use a more complete polyfill, but I don't think that it is necessary.

Yes, please use a full-featured polyfill, like the one from MDN. Reasoning is, that we might depend more on the `String.fromCodePoint` method in the future and only implementing a feature subset can then lead to harder-to-find errors. Here, I am also happy to spend a few more lines of code for the polyfill as I expect in most cases people run the code in an environment which provides `String.fromCodePoint`.

Reply:
👍

I won't copy the MDN polyfill as-is because a library shouldn't modify built-in objects. I will keep `var fromCodePoint = String.fromCodePoint || function` but with a full implementation.

Anyway, here is 2am so you will have to wait until tomorrow 😛

Reply:
SGTM about `var fromCodePoint = ...`

Thx, take your time. Sleep is important :)

PS: Out of personal interest (and as I could imagine you know the answer): Is there a common way to write the polyfill, such that the transpilers like Babel, Rollup, ... do deadcode elimination on the polyfill when the target js environment supports `String.fromCodePoint` already?

Reply:
Unfortunately there isn't 🙁. When using Babel you can avoid adding unnecessary polyfills, but only if you know the target environments (so only for apps, not for libs). Also, I don't know of other tools which reduce the code size based on the needed browser support.

Reply:
No worries. Actually, I just wondered if Closure or Prepack are smart enough to remove the polyfill? Looks like Prepack is doing a decent job here [1] and deadcode-eliminates the code.

[1] https://prepack.io/repl.html#BQMwrgdgxgLglgewgAmASmQbwFDOQI2QF5kBlGAJzggHMA6EChAWwGEEATAUwAUFqYyAD5Dk4aPCTAAhhkx5ceZBS4wwFFNOQBqZACIAjHuQB6ZACYA3IoC+1vFCQBnBABsudVwhrB8wAGwArGho1jZo6JZAA

<img width="799" alt="screen shot 2018-01-05 at 11 53 26 am" src="https://user-images.githubusercontent.com/191719/34606543-22b2808c-f210-11e7-8e90-14920fe8aec0.png">

Reply:
But since in prepack `String.fromCodePoint` is always defined, it will always remove the polyfill.

And b is not removed because you forgot `var`, so it is a global variable :stuck_out_tongue:

Reply:
> And b is not removed because you forgot var, so it is a global variable 😛

🙈 ... I was guessing this is less a Prepack bug than a jviereck-bug ^^ Thanks a lot for the explanation - that `b` is kept there as it is a global variable makes totally sense! 👍🏻👏🏻